docs/adr/0001-update-logins-api.md

```diff
@@ -0,0 +1,104 @@
+* Option 2 - Implement a different key management approach
+
+
+## Decision Outcome
+
+(I'm purposely leaving this blank until we arrive at a decision.)
+
+
+## Pros and Cons of the Options
+
+### Option 1 - Reduce the API functions that require the encryption key and pass the key to the remaining functions
+
+* Description
+
+  Currently the below logins API functions would require the per-field encryption key:
+    - [add](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L362)
+    - [update](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L611)
+    - [get_all](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L273)
+    - [get_by_base_domain](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L279)
+    - [get_by_id](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L330)
+    - [check_valid_with_no_dupes](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L663)
+    - [potential_dupes_ignoring_username](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L721)
+    - [import_multiple](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L455)
+
+      **Note:** Functions related to sync have been omitted as it is assumed they will have access to decrypted data.
+
+  Propsed changes:
+    - Combine the `add` and `update` functions into a new `add_or_update` function
+      - This will allow the removal of consumer code that distinguishes when a login record should be created or updated
+    - Pass the per-field encryption key to the `import_multiple` function
+      - This function will be removed once the Fennec to Fenix migration period ends
+    - Remove both the `potential_dupes_ignoring_username` and `check_valid_with_no_dupes` from the API
+      - Neither function is called in Firefox iOS
+      - Android Components uses both to provide validation and de-duping before logins are added or updated so we can eliminate the need to externalize these functions by replicating this logic in the new `add_or_update` function
+    - Create a `decrypt_and_fixup_login` function that both decrypts a login _and_ performs the validate and fixup logic
+      - This will eliminate the need for the `get_all`, `get_by_base_domain`, and `get_by_id` API functions to perform the fixup logic
+
+    Making the above changes will reduce the API functions requiring the encryption key to the following:
+    - `add_or_update`
```

Comment:
If we approve of combining the decryption and fixup logic and exposing that in our API, I think we should also have a similar function encryption for consistency's sake. That would mean `add_or_update` would no longer require the encryption key and a new `encrypt_and_fixup_login` would require it.

Reply:
I like the symmetry of this, which I think will be helpful for consumers to keep a coherent mental model of the operations. So on the way out of storage it would be like:

```
enc_login = store.get_by_base_domain("foo.com") // Or similar, I actually don't remember the precise API details...
if do_we_want_to_use_this(enc_login):
    key = get_the_key_from_somewhere()
    login = store.decrypt_and_fixup(key, enc_login)
    do_stuff_with_the(login)
```

And on the way in to storage it would be like:

```
enc_login = store.encrypt_and_fixup_login(key, login)
store.add_or_update(enc_login)
```

Hrm...on the other hand, it seems like the `encrypt_and_fixup_login` is going to want to do the dupe checking and what-not anyway, and from a pure implementation point-of-view it probably wants to just write the new record in there if it's not a dupe. (In fact, would there be a chance of creating dupes if two threads were running this procedure at the same time, because we do the check non-atomically with saving the data?).

Reply:
I'm liking the symmetry too and having `add_or_update` input an encrypted login would also give us the ability to update logins without ever decrypting them.  For example we could have a `last_seen` field that gets updated even if the user doesn't unlock their key.  This is bordering on YAGNI, but it still seems nice to me.

The dupe checking issue seems real to me, but it also seems like even with 1 function you would have issues.  Could we leverage the database for this?

  - Add a `dupe_hash` column that's a secure hash of the all of the fields that need to be unique together (username, hostname, realm, etc.). Add a unique constraint for this one.
  - `encypt_and_fixup_login()` calculates the hash
  - `add_or_update()` uses the hash to figure out what it should do

Reply:
This also brings up another question:  Is the login in your second snippet the same object as the login in the first?  For example, in this scenario:
  - User selects an existing login and we auto-fill the form fields
  - User changes the username field value
  - User submits the form

Do we expect and/or require the client to send us the login dict used it step 1 or will they create a new dict with login data and give us that?

Reply:
> Hrm...on the other hand, it seems like the `encrypt_and_fixup_login` is going to want to do the dupe checking and what-not anyway, and from a pure implementation point-of-view it probably wants to just write the new record in there if it's not a dupe.

Yeaaaaaah not only that, `add_or_update` not having the key means that we'd have no way of supporting `time_password_changed` because we need to compare the old and new decrypted password values before, you know, adds or updates. 

In conclusion I vote no to my own idea of adding `encrypt_and_fixup_login` to the API.

Reply:
> In conclusion I vote no to my own idea of adding encrypt_and_fixup_login to the API.

👍🏻, and let's please record the reasoning as part of this ADR because there's no way I'm going to remember it three months from now